### DISCUSSION TITLE
Moving inner classes

### DISCUSSION THREAD

**Discussion by cshjsc** [openrewrite/rewrite#4007] [Q&A] [answered]
Is there a recipe to move all the inner classes to new files?
or is there a simple way to do it?

**Comment by timtebeek** [comment, marked as answer]
Hi! No there's not yet a recipe to do so, and I suspect there would be some caveats involved with writing one that is generically applicable. 


Just off the top of my head you would need to:
1. limit yourself to static inner classes, or introduce an explicit argument for the implicit outer class argument to non-static inner class constructors
2. be mindful of library semantics like JUnit Jupiter's `@Nested` annotation on inner classes
3. update any references inside the inner class to fields and methods in the outer class

There's likely more edge cases, and some smaller items that are more easily handled, but figured outline some of the challenges first.

You're welcome to join our Slack if you want to discuss specifics with some more direct feedback. The question mostly is what you're trying to achieve and why. Starting small for a limited use case is often easier than trying to do it all in one go.
